relations/jf/src/withdraw.rs

```diff
@@ -0,0 +1,288 @@
+            token_amount_out: public.token_amount_out,
+            merkle_root: public.merkle_root,
+            merkle_proof: private.merkle_proof,
+            leaf_index: private.leaf_index,
+        }
+    }
+}
+
+impl PublicInput for WithdrawRelation {
+    fn public_input(&self) -> Vec<CircuitField> {
+        let mut public_input = vec![
+            self.fee.into(),
+            convert_account(self.recipient),
+            self.token_amount_out.into(),
+        ];
+        public_input.extend(self.spend_note.public_input());
+        public_input.extend(self.redeposit_note.public_input());
+        public_input.push(convert_array(self.merkle_root));
+
+        public_input
+    }
+}
+
+impl Relation for WithdrawRelation {
+    fn generate_subcircuit(&self, circuit: &mut PlonkCircuit<CircuitField>) -> PlonkResult<()> {
+        let _fee_var = circuit.create_public_variable(self.fee.into())?;
+        let _recipient_var = circuit.create_public_variable(convert_account(self.recipient))?;
+        let token_amount_out_var = circuit.create_public_variable(self.token_amount_out.into())?;
+        circuit.enforce_leq_constant(token_amount_out_var, CircuitField::from(u128::MAX))?;
+
+        let spend_note_var = circuit.create_note_variable(&self.spend_note)?;
+        let whole_token_amount_var = spend_note_var.token_amount_var;
+        circuit.enforce_note_preimage(spend_note_var)?;
+
+        let deposit_note_var = circuit.create_note_variable(&self.redeposit_note)?;
+        let deposit_amount_var = deposit_note_var.token_amount_var;
+        circuit.enforce_note_preimage(deposit_note_var)?;
+
+        let token_sum_var = circuit.add(token_amount_out_var, deposit_amount_var)?;
+        circuit.enforce_equal(token_sum_var, whole_token_amount_var)?;
+
+        check_merkle_proof(
+            circuit,
+            self.leaf_index,
+            self.merkle_root,
+            &self.merkle_proof,
+        )?;
+
+        Ok(())
+    }
+}
+
+// TODO refactor when implementing DepositAndMerge
+type MerkleTree = RescueSparseMerkleTree<BigUint, CircuitField>;
+type MerkleTreeGadget = dyn jf_primitives::circuit::merkle_tree::MerkleTreeGadget<
+    MerkleTree,
+    MembershipProofVar = Merkle3AryMembershipProofVar,
+    DigestGadget = RescueDigestGadget,
+>;
+type MerkleProof = <MerkleTree as MerkleTreeScheme>::MembershipProof;
+
+fn check_merkle_proof(
+    circuit: &mut PlonkCircuit<CircuitField>,
+    leaf_index: LeafIndex,
+    merkle_root: MerkleRoot,
+    merkle_proof: &MerkleProof,
+) -> PlonkResult<()> {
+    let index_var = circuit.create_variable(leaf_index.into())?;
+    let proof_var = MerkleTreeGadget::create_membership_proof_variable(circuit, merkle_proof)?;
+    let root_var = MerkleTreeGadget::create_root_variable(circuit, convert_array(merkle_root))?;
+    circuit.set_variable_public(root_var)?;
+
+    MerkleTreeGadget::enforce_membership_proof(circuit, index_var, proof_var, root_var)
+        .map_err(Into::into)
+}
+
+#[cfg(test)]
```

Comment:
I would also like to have a test for recipient/fee inclusion

Reply:
Added